### PR TITLE
Add more tokenizer tests

### LINKED ISSUE #3730: Missing tokenizer tests
AFAIU we are missing tokenizer tests for supported models like

* Baichuan
* Bloom
* GptNeoX
* Persimmon
* Refact
* Starcoder

It would be great if anyone would be helping out.

### PR DESCRIPTION
**Pull request by Galunid**
Conversion scripts used https://github.com/ggerganov/llama.cpp/commit/96981f37b1e3f450d9e63e571514217bf60f0a7f
- [X]  Baichuan (applied fix from #3739)
- [X]  Bloom
- [X]  GptNeoX
- [ ]  ~~Persimmon~~
- [X]  Refact
- [x]  Starcoder

Issues:
* Persimmon script doesn't allow for `--vocab-only`, 

* GPT-Neox tokenizer fails with `std::unordered_map` illegal access seen in other gpt2 tokenizer based models. I applied the fix from https://github.com/ggerganov/llama.cpp/issues/3730#issuecomment-1774208027 and the test passed. I didn't include the passing version here, only the failing one, let me know which one you want @goerch 
* Refact fails with `byte not found in vocab` (you can see in CI)
* Starcoder fails with `byte not found in vocab` (you can see in CI)

Models used:
- Baichuan - https://huggingface.co/baichuan-inc/Baichuan-7B
- Bloom - https://huggingface.co/bigscience/bloom-1b7/
- GptNeoX - https://huggingface.co/EleutherAI/gpt-neox-20b/ 
- Refact - https://huggingface.co/smallcloudai/Refact-1_6B-fim 
- Starcoder - https://huggingface.co/bigcode/starcoder 

closes #3730

### CODE REVIEW

**Comment by goerch** [comment]
I'm testing `refact` and `test-tokenizer-1-bpe is failing

```
    // TODO: why doesn't this work for the full range of Unicodes?
    // for (uint32_t cp = 0x10000; cp < 0x0010ffff; ++cp) {
    for (uint32_t cp = 0x10000; cp < 0x00080000; ++cp) {
[...]
```

for codepoint `0x40000`. That doesn't look like a random problem. I'm not sure if we should only check assigned Unicode [planes](https://en.wikipedia.org/wiki/Plane_(Unicode)) here (my `unicode-fu` is weak)?

**Comment by goerch** [comment]
> I'm not sure if we should only check assigned Unicode [planes](https://en.wikipedia.org/wiki/Plane_(Unicode)) here (my `unicode-fu` is weak) or if this is a real problem?

Restricting `test-tokenizer-1-bpe` to the assigned Unicode planes like so

```
    // NOTE: only testing assigned Unicode planes
    // for (uint32_t cp = 0x10000; cp < 0x0010ffff; ++cp) {
    for (uint32_t cp = 0x10000; cp < 0x00040000; ++cp) {
        std::string str = codepoint_to_utf8(cp);
        std::vector<llama_token> tokens = llama_tokenize(ctx, str, false);
        std::string check = llama_detokenize_bpe(ctx, tokens);
        if (str != check) {
            fprintf(stderr, "%s : error: codepoint %x detokenizes to '%s'(%zu) instead of '%s'(%zu)\n",
                __func__, cp, check.c_str(), check.length(), str.c_str(), str.length());
            return 4;
        }
    }
    for (uint32_t cp = 0x000e0000; cp < 0x0010ffff; ++cp) {
        std::string str = codepoint_to_utf8(cp);
        std::vector<llama_token> tokens = llama_tokenize(ctx, str, false);
        std::string check = llama_detokenize_bpe(ctx, tokens);
        if (str != check) {
            fprintf(stderr, "%s : error: codepoint %x detokenizes to '%s'(%zu) instead of '%s'(%zu)\n",
                __func__, cp, check.c_str(), check.length(), str.c_str(), str.length());
            return 4;
        }
    }
```

fixes `refact` and is working for all my models (didn't check `starcoder`).

**Comment by Galunid** [comment]
After applying your change to the test starcoder passes too
```
$ make test
Running tests...
Test project /home/kris/llama.cpp/build
      Start  1: test-quantize-fns
 1/18 Test  #1: test-quantize-fns ................   Passed    0.02 sec
      Start  2: test-quantize-perf
 2/18 Test  #2: test-quantize-perf ...............   Passed    0.07 sec
      Start  3: test-sampling
 3/18 Test  #3: test-sampling ....................   Passed    0.01 sec
      Start  4: test-tokenizer-0-llama
 4/18 Test  #4: test-tokenizer-0-llama ...........   Passed    0.07 sec
      Start  5: test-tokenizer-0-falcon
 5/18 Test  #5: test-tokenizer-0-falcon ..........   Passed    0.29 sec
      Start  6: test-tokenizer-1-llama
 6/18 Test  #6: test-tokenizer-1-llama ...........   Passed    2.88 sec
      Start  7: test-tokenizer-1-baichuan
 7/18 Test  #7: test-tokenizer-1-baichuan ........   Passed    3.09 sec
      Start  8: test-tokenizer-1-falcon
 8/18 Test  #8: test-tokenizer-1-falcon ..........   Passed    1.88 sec
      Start  9: test-tokenizer-1-aquila
 9/18 Test  #9: test-tokenizer-1-aquila ..........   Passed    1.95 sec
      Start 10: test-tokenizer-1-mpt
10/18 Test #10: test-tokenizer-1-mpt .............   Passed    1.33 sec
      Start 11: test-tokenizer-1-bloom
11/18 Test #11: test-tokenizer-1-bloom ...........   Passed    4.56 sec
      Start 12: test-tokenizer-1-gpt-neox
12/18 Test #12: test-tokenizer-1-gpt-neox ........Subprocess aborted***Exception:   0.91 sec
      Start 13: test-tokenizer-1-refact
13/18 Test #13: test-tokenizer-1-refact ..........   Passed    1.24 sec
      Start 14: test-tokenizer-1-starcoder
14/18 Test #14: test-tokenizer-1-starcoder .......   Passed    1.26 sec
      Start 15: test-grammar-parser
15/18 Test #15: test-grammar-parser ..............   Passed    0.00 sec
      Start 16: test-llama-grammar
16/18 Test #16: test-llama-grammar ...............   Passed    0.00 sec
      Start 17: test-grad0
17/18 Test #17: test-grad0 .......................   Passed    4.41 sec
      Start 18: test-rope
18/18 Test #18: test-rope ........................   Passed    0.04 sec

94% tests passed, 1 tests failed out of 18

Total Test time (real) =  24.01 sec
```

**Comment by goerch** [comment]
> Start 12: test-tokenizer-1-gpt-neox
> 12/18 Test #12: test-tokenizer-1-gpt-neox ........Subprocess aborted***Exception:   0.91 sec

Interesting. I see

```
      Start 12: test-tokenizer-1-gpt-neox
12/17 Test #12: test-tokenizer-1-gpt-neox ........   Passed    1.87 sec
```

Hm.

**Comment by Galunid** [comment]
Pay no attention to neox, it's not relevant here (uses old model that was failing on map illegal access)

**Comment by Galunid** [comment]
Should I also include your patch to restrict tokenizer tests to unicode planes in this PR?

File: tests/test-tokenizer-1-bpe.cpp [hunk 1]
```diff
@@ -93,7 +93,17 @@ int main(int argc, char **argv) {
     }
     // TODO: why doesn't this work for the full range of Unicodes?
```
Comment (goerch) [inline tests/test-tokenizer-1-bpe.cpp:94]: Could be helpful to change the comment?

Comment (Galunid) [reply tests/test-tokenizer-1-bpe.cpp:94]: Does `Restrict to unicode planes` sound good?

**Review by goerch — APPROVED** [review]
Nicely done!

**Comment by goerch** [comment]
@Galunid : sorry for mixing it up when using Github online editor. May I ask you to repair the mess (don't know enough about git to fix this on your branch)? Thanks again.

File: models/ggml-vocab-bloom.gguf
Comment (ggerganov) [inline models/ggml-vocab-bloom.gguf:1]: The bloom vocab is 12MB - this is too much. Let's remove it

Comment (goerch) [reply models/ggml-vocab-bloom.gguf:1]: Please (out-)comment the test case with `#BIG` (and remove the binary). At least you tried it :)

Comment (Galunid) [reply models/ggml-vocab-bloom.gguf:1]: Thanks, good catch!

**Comment by goerch** [comment]
OK, I'll wait until tomorrow morning (+10 hours) before merging.

**Pull request merged**